README.md
[![Build Status](https://travis-ci.com/daveoncode/python-string-utils.svg?branch=develop)](https://travis-ci.com/daveoncode/python-string-utils)
[![codecov](https://codecov.io/gh/daveoncode/python-string-utils/branch/develop/graph/badge.svg)](https://codecov.io/gh/daveoncode/python-string-utils)
[![Documentation Status](https://readthedocs.org/projects/python-string-utils/badge/?version=develop)](https://python-string-utils.readthedocs.io/en/develop)
[![Python 3.5](https://img.shields.io/badge/python-3.5-blue.svg)](https://www.python.org/downloads/release/python-350/)
[![Python 3.6](https://img.shields.io/badge/python-3.6-blue.svg)](https://www.python.org/downloads/release/python-360/)
[![Python 3.7](https://img.shields.io/badge/python-3.7-blue.svg)](https://www.python.org/downloads/release/python-370/)
[![Python 3.8](https://img.shields.io/badge/python-3.8-blue.svg)](https://www.python.org/downloads/release/python-380/)

# Python String Utils
##### Latest version: [1.0.0](https://github.com/daveoncode/python-string-utils/blob/master/CHANGELOG.md) (March 2020)

A handy library to validate, manipulate and generate strings, which is:

- Simple and "pythonic"
- Fully documented and with examples! (html version on [readthedocs.io](https://python-string-utils.readthedocs.io/en/develop))
- 100% code coverage! (see it with your own eyes on [codecov.io](https://codecov.io/gh/daveoncode/python-string-utils/branch/develop))
- Tested (automatically on each push thanks to Travis CI) against all officially supported Python versions
- Fast (mostly based on compiled regex)
- Free from external dependencies
- PEP8 complaint


## What's inside...

### Library structure

The library basically consists in the python package `string_utils`, containing the following modules:

- `validation.py` (contains string check api)
- `manipulation.py` (contains string transformation api)
- `generation.py` (contains string generation api)
- `errors.py` (contains library-specific errors)
- `_regex.py` (contains compiled regex **FOR INTERNAL USAGE ONLY**)

Plus a secondary package `tests` which includes several submodules.\
Specifically one for each test suite and named according to the api to test (eg. tests for `is_ip()` 
will be in `test_is_ip.py` and so on).

All the public API are importable directly from the main package `string_utils`, so this:

`from string_utils.validation import is_ip`

can be simplified as:
 
 `from string_utils import is_ip`

### Api overview

***Bear in mind**: this is just an overview, 
for the full API documentation see:* [readthedocs.io](https://python-string-utils.readthedocs.io/en/develop)

#### String validation functions:


**is_string**: checks if the given object is a string
~~~~
is_string('hello') # returns true
is_string(b'hello') # returns false
~~~~

**is_full_string**: checks if the given object is non empty string
~~~~
is_full_string(None) # returns false
is_full_string('') # returns false
is_full_string(' ') # returns false
is_full_string('foo') # returns true
~~~~

**is_number**: checks if the given string represents a valid number
~~~~
is_number('42') # returns true
is_number('-25.99') # returns true
is_number('1e3') # returns true
is_number(' 1 2 3 ') # returns false
~~~~

**is_integer**: checks if the given string represents a valid integer
~~~~
is_integer('42') # returns true
is_integer('42.0') # returns false
~~~~

**is_decimal**: checks if the given string represents a valid decimal number
~~~~
is_decimal('42.0') # returns true
is_decimal('42') # returns false
~~~~

**is_url**: checks if the given string is an url
~~~~
is_url('foo.com') # returns false
is_url('http://www.foo.com') # returns true
is_url('https://foo.com') # returns true
~~~~

**is_email**: Checks if the given string is an email
~~~~
is_email('my.email@some.provider.com') # returns true
is_eamil('@gmail.com') # retruns false
~~~~

**is_credit_card**: Checks if the given string is a credit card
~~~~
is_credit_card(value)

# returns true if `value` represents a valid card number for one of these:
# VISA, MASTERCARD, AMERICAN EXPRESS, DINERS CLUB, DISCOVER or JCB
~~~~

**is_camel_case**: Checks if the given string is formatted as camel case
~~~~
is_camel_case('MyCamelCase') # returns true
is_camel_case('hello') # returns false
~~~~

**is_snake_case**: Checks if the given string is formatted as snake case
~~~~
is_snake_case('snake_bites') # returns true
is_snake_case('nope') # returns false
~~~~

**is_json**: Checks if the given string is a valid json
~~~~
is_json('{"first_name": "Peter", "last_name": "Parker"}') # returns true
is_json('[1, 2, 3]') # returns true
is_json('{nope}') # returns false
~~~~

**is_uuid**: Checks if the given string is a valid UUID
~~~~
is_uuid('ce2cd4ee-83de-46f6-a054-5ee4ddae1582') # returns true
~~~~

**is_ip_v4**: Checks if the given string is a valid ip v4 address
~~~~
is_ip_v4('255.200.100.75') # returns true
is_ip_v4('255.200.100.999') # returns false (999 is out of range)
~~~~

**is_ip_v6**: Checks if the given string is a valid ip v6 address
~~~~
is_ip_v6('2001:db8:85a3:0000:0000:8a2e:370:7334') # returns true
is_ip_v6('123:db8:85a3:0000:0000:8a2e:370,1') # returns false
~~~~

**is_ip**: Checks if the given string is a valid ip (any version)
~~~~
is_ip('255.200.100.75') # returns true
is_ip('2001:db8:85a3:0000:0000:8a2e:370:7334') # returns true
is_ip('255.200.100.999') # returns false
is_ip('123:db8:85a3:0000:0000:8a2e:370,1') # returns false
~~~~

**is_isnb_13**: Checks if the given string is a valid ISBN 13
~~~~
is_isbn_13('9780312498580') # returns true
is_isbn_13('978-0312498580') # returns true
is_isbn_13('978-0312498580', normalize=False) # returns false
~~~~

**is_isbn_10**: Checks if the given string is a valid ISBN 10
~~~~
is_isbn_10('1506715214') # returns true
is_isbn_10('150-6715214') # returns true
is_isbn_10('150-6715214', normalize=False) # returns false
~~~~

**is_isbn**: Checks if the given string is a valid ISBN (any version)
~~~~
is_isbn('9780312498580') # returns true
is_isbn('1506715214') # returns true
~~~~

**is_slug**: Checks if the string is a slug (as created by `slugify()`)
~~~~
is_slug('my-blog-post-title') # returns true
is_slug('My blog post title') # returns false
~~~~

**contains_html**: Checks if the strings contains one ore more HTML/XML tag
~~~~
contains_html('my string is <strong>bold</strong>') # returns true
contains_html('my string is not bold') # returns false
~~~~

**words_count**: Returns the number of words contained in the string
~~~~
words_count('hello world') # returns 2
words_count('one,two,three') # returns 3 (no need for spaces, punctuation is recognized!)
~~~~

**is_palindrome**: Checks if the string is a palindrome
~~~~
is_palindrome('LOL') # returns true
is_palindrome('ROTFL') # returns false
~~~~

**is_pangram**: Checks if the string is a pangram
~~~~
is_pangram('The quick brown fox jumps over the lazy dog') # returns true
is_pangram('hello world') # returns false
~~~~

**is_isogram**: Checks if the string is an isogram
~~~~
is_isogram('dermatoglyphics') # returns true
is_isogram('hello') # returns false
~~~~

#### String manipulation:

**camel_case_to_snake**:  Converts a camel case formatted string into a snake case one
~~~~
camel_case_to_snake('ThisIsACamelStringTest') # returns 'this_is_a_camel_case_string_test'
~~~~

**snake_case_to_camel**: Converts a snake case formatted string into a camel case one
~~~~
snake_case_to_camel('the_snake_is_green') # returns 'TheSnakeIsGreen'
~~~~

**reverse**: Returns the string in a reversed order
~~~
reverse('hello') # returns 'olleh'
~~~

**shuffle**: Returns the string with its original chars but at randomized positions
~~~~
shuffle('hello world') # possible output: 'l wodheorll'
~~~~

**strip_html**: Removes all the HTML/XML tags found in a string
~~~~
strip_html('test: <a href="foo/bar">click here</a>') # returns 'test: '
strip_html('test: <a href="foo/bar">click here</a>', keep_tag_content=True) # returns 'test: click here'
~~~~

**prettify**: Reformat a string by applying basic grammar and formatting rules
~~~~
prettify(' unprettified string ,, like this one,will be"prettified" .it\' s awesome! ')
# the ouput will be: 'Unprettified string, like this one, will be "prettified". It\'s awesome!'
~~~~

**asciify**: Converts all non-ascii chars contained in a string into the closest possible ascii representation
~~~~
asciify('èéùúòóäåëýñÅÀÁÇÌÍÑÓË') 
# returns 'eeuuooaaeynAAACIINOE' (string is deliberately dumb in order to show char conversion)
~~~~

**slugify**: Convert a string into a formatted "slug"
~~~~
slugify('Top 10 Reasons To Love Dogs!!!') # returns: 'top-10-reasons-to-love-dogs'
~~~~

**booleanize**: Convert a string into a boolean based on its content
~~~~
booleanize('true') # returns true
booleanize('YES') # returns true
booleanize('y') # returns true
booleanize('1') # returns true
booelanize('something else') # returns false
~~~~

**strip_margin**: Removes left indentation from multi-line strings (inspired by Scala)
~~~~
strip_margin('''
        line 1
        line 2
        line 3
''')

#returns:
'''
line 1
line 2
line 3
'''
~~~~

**compress/decompress**: Compress strings into shorted ones that can be restored back to the original one later on
~~~~
compressed = compress(my_long_string) # shorter string (URL safe base64 encoded)

decompressed = decompress(compressed) # string restored

assert(my_long_string == decompressed) # yep
~~~~

**roman_encode**: Encode integers/string into roman numbers
~~~
roman_encode(37) # returns 'XXXVII'
~~~~

**roman_decode**: Decode roman number into an integer
~~~~
roman_decode('XXXVII') # returns 37
~~~~

**roman_range**: Generator which returns roman numbers on each iteration
~~~~
for n in roman_range(10): print(n) # prints: I, II, III, IV, V, VI, VII, VIII, IX, X
for n in roman_range(start=7, stop=1, step=-1): print(n) # prints: VII, VI, V, IV, III, II, I
~~~~

#### String generations:

**uuid**: Returns the string representation of a newly created UUID object
~~~~
uuid() # possible output: 'ce2cd4ee-83de-46f6-a054-5ee4ddae1582'
uuid(as_hex=True) # possible output: 'ce2cd4ee83de46f6a0545ee4ddae1582'
~~~~

**random_string**: Creates a string of the specified size with random chars
~~~~
random_string(9) # possible output: 'K1URtlTu5'
~~~~

**secure_random_hex**: Creates an hexadecimal string using a secure strong random generator
~~~~
secure_random_hex(12) 
# possible ouput: 'd1eedff4033a2e9867c37ded' 
# (len is 24, because 12 represents the number of random bytes generated, which are then converted to hexadecimal value)

CHANGELOG.md
# Python String Utils Changelog

## v1.0.0 (March 2020)

### Project reboot:

I've started this project back in 2015, then abandoned it one year later, thinking no one care about it.\
Recently I've instead discovered that it's actually being used by several projects on GitHub 
(despite it has only few stars).\
So I'm rebooting it with new found enthusiasm and a brand new version with tons of new api and improvements.

Besides the code, the project configuration is now mature and professional: 
- Automatic builds and testing against multiple python versions with Travis CI
- Automatic code coverage reporting on codecov.io
- Automatic documentation generated on readthedocs.io
- Better descriptions and provided examples


### Deprecations:

We are in 2020 and finally old versions of Python have been 
[officially deprecated](https://www.python.org/doc/sunset-python-2/).
So the following versions are **no longer supported**:

- 2.7.x
- 3.3.x
- 3.4.x

(suite tests are now being executed against all currently supported versions: 3.5, 3.6, 3.7 and 3.8)

### Added:

- `compress()`: compress strings into shorter ones that can be restored later on
- `decompress()`: restore a previously compressed string
- `roman_encode()`: encode integers/strings into roman number strings
- `roman_decode()`: decode roman number into an integer
- `roman_range()`: generator which returns roman numbers on each iteration
- `asciify()`: Force string content to be ascii-only by translating all non-ascii chars into the closest possible
 representation 
- `is_ip_v4()`: checks only for v4 ips
- `is_ip_v6()`: checks only for v6 ips
- `is_isbn_13()`: checks if the given string is a valid ISBN 13
- `is_isbn_10()`: checks if the given string is a valid ISBN 10
- `is_isbn()`: checks if the given string is a valid ISBN (any version)
- `is_number()`: checks if the given string is a valid number (either an integer or a decimal)
- `is_integer()`: checks if the given string is a valid integer
- `is_decimal()`: checks if the given string is a valid decimal
- `booleanize()`: turns string into boolean based on its content
- `strip_margin()`: remove left margin from multi line strings so you don't have to bother about indentation 
in your code (inspired by Scala)
- `random_string()`: generates string of given size with random alpha-numeric chars
- `secure_random_hex()`: generates hexadecimal string of the given bytes count using secure random generator

### Fixes:

- `is_email()` is now fully compliant with email specifications (https://tools.ietf.org/html/rfc3696#section-3)
- `is_ip()` now checks both ip v4 (and validates 0-255 range) and ip v6 
(the previous implementation was really shallow, my apologies :P)
- `is_json()` now considers as valid json array objects (eg. `is_json('[1, 2, 3]')` returns true now)
- `prettify()` does not screw up urls or emails anymore (from now on it won't consider those as text to be formatted)
- Solved deprecation warnings over invalid escape sequences in Python >= 3.7 

### Changes:

- Old module `string_utils.py` has been replaced by a package with submodules (`validation.py`, `manipulation.py`, 
`generation.py` and `errors.py`), anyway all the functions are still 
importable as before (`from string_utils import xxx`). Similarly `tests.py` has been refactored into a package 
with a module for each test case
- `is_snake_case()` now considers as "snake case" strings with mixed upper and lower case characters, strings with 
leading or trailing underscores and string containing multiple underscores in sequence
- `is_slug()` now allows multiple consecutive separator signs between words

### Improvements:

- Added Python type hints to all functions arguments and return types 
(this is now feasible since the minimum supported Python version is the 3.5)
- Each method that expect a valid string as input now will raise a more detailed `InvalidInputError` exception 
(eg: ***Expected "str", received "list"***)
- `reverse()`, `shuffle()`, `prettify()` now raise the detailed `InvalidInputError` if input is not a valid string
- String checks should now be a bit faster (switched from `.search()` to `.match()` in internal regex when the goal 
is to match the full string)
- `is_palindrome()` algorithm has been redesigned to offer a faster check and better memory usage 
(only 2 chars are now being access at the same time instead of the whole string)... 
signature has changed (now it has two optional boolean arguments: `ignore_spaces` and `ignore_case`)
- `slugify()` is now able to translate more non-ascii chars during the string conversion 
(it now makes use of the new extracted method `asciify()`)
- `is_uuid()` has now a second parameter `allow_hex` that if true, considers as valid UUID hex value
- `uuid()` has now an optional boolean parameter `as_hex` which allows to return UUID string as hex representation
- `shuffle()` is now faster

---



## v0.6.0

### Added:

- slugify
- is_slug

## v0.5.0

### Added:

- is_full_string
- is_isogram

## v0.4.2

### Fixed:

- is_url

## v0.4.1

### Changed:

- is_palindrome and is_pangram now return False if the given object is not a string instead of raising an exception

## v0.4.0

### Added:

- is_palindrome
- is_pangram

### Changed:

- reverse (improved implementation)


string_utils/validation.py
# -*- coding: utf-8 -*-

import json
import string
from typing import Any, Optional, List

__all__ = [
    'is_string',
    'is_full_string',
    'is_number',
    'is_integer',
    'is_decimal',
    'is_url',
    'is_email',
    'is_credit_card',
    'is_camel_case',
    'is_snake_case',
    'is_json',
    'is_uuid',
    'is_ip_v4',
    'is_ip_v6',
    'is_ip',
    'is_isbn_10',
    'is_isbn_13',
    'is_isbn',
    'is_palindrome',
    'is_pangram',
    'is_isogram',
    'is_slug',
    'contains_html',
    'words_count',
]

from ._regex import *
from string_utils.errors import InvalidInputError


# PRIVATE API


class _ISBNChecker:
    def __init__(self, input_string: str, normalize: bool = True):
        if not is_string(input_string):
            raise InvalidInputError(input_string)

        self.input_string = input_string.replace('-', '') if normalize else input_string

    def is_isbn_13(self) -> bool:
        if len(self.input_string) == 13:
            product = 0

            try:
                for index, digit in enumerate(self.input_string):
                    weight = 1 if (index % 2 == 0) else 3
                    product += int(digit) * weight

                return product % 10 == 0

            except ValueError:
                pass

        return False

    def is_isbn_10(self) -> bool:
        if len(self.input_string) == 10:
            product = 0

            try:
                for index, digit in enumerate(self.input_string):
                    product += int(digit) * (index + 1)

                return product % 11 == 0

            except ValueError:
                pass

        return False


# PUBLIC API

def is_string(obj: Any) -> bool:
    """
    Checks if an object is a string.

    *Example:*

    >>> is_string('foo') # returns true
    >>> is_string(b'foo') # returns false

    :param obj: Object to test.
    :return: True if string, false otherwise.
    """
    return isinstance(obj, str)


def is_full_string(input_string: Any) -> bool:
    """
    Check if a string is not empty (it must contains at least one non space character).

    *Examples:*

    >>> is_full_string(None) # returns false
    >>> is_full_string('') # returns false
    >>> is_full_string(' ') # returns false
    >>> is_full_string('hello') # returns true

    :param input_string: String to check.
    :type input_string: str
    :return: True if not empty, false otherwise.
    """
    return is_string(input_string) and input_string.strip() != ''


def is_number(input_string: str) -> bool:
    """
    Checks if a string is a valid number.

    The number can be a signed (eg: +1, -2, -3.3) or unsigned (eg: 1, 2, 3.3) integer or double
    or use the "scientific notation" (eg: 1e5).

    *Examples:*

    >>> is_number('42') # returns true
    >>> is_number('19.99') # returns true
    >>> is_number('-9.12') # returns true
    >>> is_number('1e3') # returns true
    >>> is_number('1 2 3') # returns false

    :param input_string: String to check
    :type input_string: str
    :return: True if the string represents a number, false otherwise
    """
    if not isinstance(input_string, str):
        raise InvalidInputError(input_string)

    return NUMBER_RE.match(input_string) is not None


def is_integer(input_string: str) -> bool:
    """
    Checks whether the given string represents an integer or not.

    An integer may be signed or unsigned or use a "scientific notation".

    *Examples:*

    >>> is_integer('42') # returns true
    >>> is_integer('42.0') # returns false

    :param input_string: String to check
    :type input_string: str
    :return: True if integer, false otherwise
    """
    return is_number(input_string) and '.' not in input_string


def is_decimal(input_string: str) -> bool:
    """
    Checks whether the given string represents a decimal or not.

    A decimal may be signed or unsigned or use a "scientific notation".

    >>> is_decimal('42.0') # returns true
    >>> is_decimal('42') # returns false

    :param input_string: String to check
    :type input_string: str
    :return: True if integer, false otherwise
    """
    return is_number(input_string) and '.' in input_string


# Full url example:
# scheme://username:password@www.domain.com:8042/folder/subfolder/file.extension?param=value&param2=value2#hash
def is_url(input_string: Any, allowed_schemes: Optional[List[str]] = None) -> bool:
    """
    Check if a string is a valid url.

    *Examples:*

    >>> is_url('http://www.mysite.com') # returns true
    >>> is_url('https://mysite.com') # returns true
    >>> is_url('.mysite.com') # returns false

    :param input_string: String to check.
    :type input_string: str
    :param allowed_schemes: List of valid schemes ('http', 'https', 'ftp'...). Default to None (any scheme is valid).
    :type allowed_schemes: Optional[List[str]]
    :return: True if url, false otherwise
    """
    if not is_full_string(input_string):
        return False

    valid = URL_RE.match(input_string) is not None

    if allowed_schemes:
        return valid and any([input_string.startswith(s) for s in allowed_schemes])

    return valid


def is_email(input_string: Any) -> bool:
    """
    Check if a string is a valid email.

    Reference: https://tools.ietf.org/html/rfc3696#section-3

    *Examples:*

    >>> is_email('my.email@the-provider.com') # returns true
    >>> is_email('@gmail.com') # returns false

    :param input_string: String to check.
    :type input_string: str
    :return: True if email, false otherwise.
    """
    # first simple "pre check": it must be a non empty string with max len 320 and cannot start with a dot
    if not is_full_string(input_string) or len(input_string) > 320 or input_string.startswith('.'):
        return False

    try:
        # we expect 2 tokens, one before "@" and one after, otherwise we have an exception and the email is not valid
        head, tail = input_string.split('@')

        # head's size must be <= 64, tail <= 255, head must not start with a dot or contain multiple consecutive dots
        if len(head) > 64 or len(tail) > 255 or head.endswith('.') or ('..' in head):
            return False

        # removes escaped spaces, so that later on the test regex will accept the string
        head = head.replace('\\ ', '')
        if head.startswith('"') and head.endswith('"'):
            head = head.replace(' ', '')[1:-1]

        return EMAIL_RE.match(head + '@' + tail) is not None

    except ValueError:
        # borderline case in which we have multiple "@" signs but the head part is correctly escaped
        if ESCAPED_AT_SIGN.search(input_string) is not None:
            # replace "@" with "a" in the head
            return is_email(ESCAPED_AT_SIGN.sub('a', input_string))

        return False


def is_credit_card(input_string: Any, card_type: str = None) -> bool:
    """
    Checks if a string is a valid credit card number.
    If card type is provided then it checks against that specific type only,
    otherwise any known credit card number will be accepted.

    Supported card types are the following:

    - VISA
    - MASTERCARD
    - AMERICAN_EXPRESS
    - DINERS_CLUB
    - DISCOVER
    - JCB

    :param input_string: String to check.
    :type input_string: str
    :param card_type: Card type. Default to None (any card).
    :type card_type: str

    :return: True if credit card, false otherwise.
    """
    if not is_full_string(input_string):
        return False

    if card_type:
        if card_type not in CREDIT_CARDS:
            raise KeyError(
                'Invalid card type "{}". Valid types are: {}'.format(card_type, ', '.join(CREDIT_CARDS.keys()))
            )
        return CREDIT_CARDS[card_type].match(input_string) is not None

    for c in CREDIT_CARDS:
        if CREDIT_CARDS[c].match(input_string) is not None:
            return True

    return False


def is_camel_case(input_string: Any) -> bool:
    """
    Checks if a string is formatted as camel case.

    A string is considered camel case when:

    - it's composed only by letters ([a-zA-Z]) and optionally numbers ([0-9])
    - it contains both lowercase and uppercase letters
    - it does not start with a number

    *Examples:*

    >>> is_camel_case('MyString') # returns true
    >>> is_camel_case('mystring') # returns false

    :param input_string: String to test.
    :type input_string: str
    :return: True for a camel case string, false otherwise.
    """
    return is_full_string(input_string) and CAMEL_CASE_TEST_RE.match(input_string) is not None


def is_snake_case(input_string: Any, separator: str = '_') -> bool:
    """
    Checks if a string is formatted as "snake case".

    A string is considered snake case when:

    - it's composed only by lowercase/uppercase letters and digits
    - it contains at least one underscore (or provided separator)
    - it does not start with a number

    *Examples:*

    >>> is_snake_case('foo_bar_baz') # returns true
    >>> is_snake_case('foo') # returns false

    :param input_string: String to test.
    :type input_string: str
    :param separator: String to use as separator.
    :type separator: str
    :return: True for a snake case string, false otherwise.
    """
    if is_full_string(input_string):
        re_map = {
            '_': SNAKE_CASE_TEST_RE,
            '-': SNAKE_CASE_TEST_DASH_RE
        }
        re_template = r'([a-z]+\d*{sign}[a-z\d{sign}]*|{sign}+[a-z\d]+[a-z\d{sign}]*)'
        r = re_map.get(
            separator,
            re.compile(re_template.format(sign=re.escape(separator)), re.IGNORECASE)
        )

        return r.match(input_string) is not None

    return False


def is_json(input_string: Any) -> bool:
    """
    Check if a string is a valid json.

    *Examples:*

    >>> is_json('{"name": "Peter"}') # returns true
    >>> is_json('[1, 2, 3]') # returns true
    >>> is_json('{nope}') # returns false

    :param input_string: String to check.
    :type input_string: str
    :return: True if json, false otherwise
    """
    if is_full_string(input_string) and JSON_WRAPPER_RE.match(input_string) is not None:
        try:
            return isinstance(json.loads(input_string), (dict, list))
        except (TypeError, ValueError, OverflowError):
            pass

    return False


def is_uuid(input_string: Any, allow_hex: bool = False) -> bool:
    """
    Check if a string is a valid UUID.

    *Example:*

    >>> is_uuid('6f8aa2f9-686c-4ac3-8766-5712354a04cf') # returns true
    >>> is_uuid('6f8aa2f9686c4ac387665712354a04cf') # returns false
    >>> is_uuid('6f8aa2f9686c4ac387665712354a04cf', allow_hex=True) # returns true

    :param input_string: String to check.
    :type input_string: str
    :param allow_hex: True to allow UUID hex representation as valid, false otherwise (default)
    :type allow_hex: bool
    :return: True if UUID, false otherwise
    """
    # string casting is used to allow UUID itself as input data type
    s = str(input_string)

    if allow_hex:
        return UUID_HEX_OK_RE.match(s) is not None

    return UUID_RE.match(s) is not None


def is_ip_v4(input_string: Any) -> bool:
    """
    Checks if a string is a valid ip v4.

    *Examples:*

    >>> is_ip_v4('255.200.100.75') # returns true
    >>> is_ip_v4('nope') # returns false (not an ip)
    >>> is_ip_v4('255.200.100.999') # returns false (999 is out of range)

    :param input_string: String to check.
    :type input_string: str
    :return: True if an ip v4, false otherwise.
    """
    if not is_full_string(input_string) or SHALLOW_IP_V4_RE.match(input_string) is None:
        return False

    # checks that each entry in the ip is in the valid range (0 to 255)
    for token in input_string.split('.'):
        if not (0 <= int(token) <= 255):
            return False

    return True


def is_ip_v6(input_string: Any) -> bool:
    """
    Checks if a string is a valid ip v6.

    *Examples:*

    >>> is_ip_v6('2001:db8:85a3:0000:0000:8a2e:370:7334') # returns true
    >>> is_ip_v6('2001:db8:85a3:0000:0000:8a2e:370:?') # returns false (invalid "?")

    :param input_string: String to check.
    :type input_string: str
    :return: True if a v6 ip, false otherwise.
    """
    return is_full_string(input_string) and IP_V6_RE.match(input_string) is not None


def is_ip(input_string: Any) -> bool:
    """
    Checks if a string is a valid ip (either v4 or v6).

    *Examples:*

    >>> is_ip('255.200.100.75') # returns true
    >>> is_ip('2001:db8:85a3:0000:0000:8a2e:370:7334') # returns true
    >>> is_ip('1.2.3') # returns false

    :param input_string: String to check.
    :type input_string: str
    :return: True if an ip, false otherwise.
    """
    return is_ip_v6(input_string) or is_ip_v4(input_string)


def is_palindrome(input_string: Any, ignore_spaces: bool = False, ignore_case: bool = False) -> bool:
    """
    Checks if the string is a palindrome (https://en.wikipedia.org/wiki/Palindrome).

    *Examples:*

    >>> is_palindrome('LOL') # returns true
    >>> is_palindrome('Lol') # returns false
    >>> is_palindrome('Lol', ignore_case=True) # returns true
    >>> is_palindrome('ROTFL') # returns false

    :param input_string: String to check.
    :type input_string: str
    :param ignore_spaces: False if white spaces matter (default), true otherwise.
    :type ignore_spaces: bool
    :param ignore_case: False if char case matters (default), true otherwise.
    :type ignore_case: bool
    :return: True if the string is a palindrome (like "otto", or "i topi non avevano nipoti" if strict=False),\
    False otherwise
    """
    if not is_full_string(input_string):
        return False

    if ignore_spaces:
        input_string = SPACES_RE.sub('', input_string)

    string_len = len(input_string)

    # Traverse the string one char at step, and for each step compares the
    # "head_char" (the one on the left of the string) to the "tail_char" (the one on the right).
    # In this way we avoid to manipulate the whole string in advance if not necessary and provide a faster
    # algorithm which can scale very well for long strings.
    for index in range(string_len):
        head_char = input_string[index]
        tail_char = input_string[string_len - index - 1]

        if ignore_case:
            head_char = head_char.lower()
            tail_char = tail_char.lower()

        if head_char != tail_char:
            return False

    return True


def is_pangram(input_string: Any) -> bool:
    """
    Checks if the string is a pangram (https://en.wikipedia.org/wiki/Pangram).

    *Examples:*

    >>> is_pangram('The quick brown fox jumps over the lazy dog') # returns true
    >>> is_pangram('hello world') # returns false

    :param input_string: String to check.
    :type input_string: str
    :return: True if the string is a pangram, False otherwise.
    """
    if not is_full_string(input_string):
        return False

    return set(SPACES_RE.sub('', input_string)).issuperset(set(string.ascii_lowercase))


def is_isogram(input_string: Any) -> bool:
    """
    Checks if the string is an isogram (https://en.wikipedia.org/wiki/Isogram).

    *Examples:*

    >>> is_isogram('dermatoglyphics') # returns true
    >>> is_isogram('hello') # returns false

    :param input_string: String to check.
    :type input_string: str
    :return: True if isogram, false otherwise.
    """
    return is_full_string(input_string) and len(set(input_string)) == len(input_string)


def is_slug(input_string: Any, separator: str = '-') -> bool:
    """
    Checks if a given string is a slug (as created by `slugify()`).

    *Examples:*

    >>> is_slug('my-blog-post-title') # returns true
    >>> is_slug('My blog post title') # returns false

    :param input_string: String to check.
    :type input_string: str
    :param separator: Join sign used by the slug.
    :type separator: str
    :return: True if slug, false otherwise.
    """
    if not is_full_string(input_string):
        return False

    rex = r'^([a-z\d]+' + re.escape(separator) + r'*?)*[a-z\d]$'

    return re.match(rex, input_string) is not None


def contains_html(input_string: str) -> bool:
    """
    Checks if the given string contains HTML/XML tags.

    By design, this function matches ANY type of tag, so don't expect to use it
    as an HTML validator, its goal is to detect "malicious" or undesired tags in the text.

    *Examples:*

    >>> contains_html('my string is <strong>bold</strong>') # returns true
    >>> contains_html('my string is not bold') # returns false

    :param input_string: Text to check
    :type input_string: str
    :return: True if string contains html, false otherwise.
    """
    if not is_string(input_string):
        raise InvalidInputError(input_string)

    return HTML_RE.search(input_string) is not None


def words_count(input_string: str) -> int:
    """
    Returns the number of words contained into the given string.

    This method is smart, it does consider only sequence of one or more letter and/or numbers
    as "words", so a string like this: "! @ # % ... []" will return zero!
    Moreover it is aware of punctuation, so the count for a string like "one,two,three.stop"
    will be 4 not 1 (even if there are no spaces in the string).

    *Examples:*

    >>> words_count('hello world') # returns 2
    >>> words_count('one,two,three.stop') # returns 4

    :param input_string: String to check.
    :type input_string: str
    :return: Number of words.
    """
    if not is_string(input_string):
        raise InvalidInputError(input_string)

    return len(WORDS_COUNT_RE.findall(input_string))


def is_isbn_10(input_string: str, normalize: bool = True) -> bool:
    """
    Checks if the given string represents a valid ISBN 10 (International Standard Book Number).
    By default hyphens in the string are ignored, so digits can be separated in different ways, by calling this
    function with `normalize=False` only digit-only strings will pass the validation.

    *Examples:*

    >>> is_isbn_10('1506715214') # returns true
    >>> is_isbn_10('150-6715214') # returns true
    >>> is_isbn_10('150-6715214', normalize=False) # returns false

    :param input_string: String to check.
    :param normalize: True to ignore hyphens ("-") in the string (default), false otherwise.
    :return: True if valid ISBN 10, false otherwise.
    """
    checker = _ISBNChecker(input_string, normalize)
    return checker.is_isbn_10()


def is_isbn_13(input_string: str, normalize: bool = True) -> bool:
    """
    Checks if the given string represents a valid ISBN 13 (International Standard Book Number).
    By default hyphens in the string are ignored, so digits can be separated in different ways, by calling this
    function with `normalize=False` only digit-only strings will pass the validation.

    *Examples:*

    >>> is_isbn_13('9780312498580') # returns true
    >>> is_isbn_13('978-0312498580') # returns true
    >>> is_isbn_13('978-0312498580', normalize=False) # returns false

    :param input_string: String to check.
    :param normalize: True to ignore hyphens ("-") in the string (default), false otherwise.
    :return: True if valid ISBN 13, false otherwise.
    """
    checker = _ISBNChecker(input_string, normalize)
    return checker.is_isbn_13()


def is_isbn(input_string: str, normalize: bool = True) -> bool:
    """
    Checks if the given string represents a valid ISBN (International Standard Book Number).
    By default hyphens in the string are ignored, so digits can be separated in different ways, by calling this
    function with `normalize=False` only digit-only strings will pass the validation.

    *Examples:*

    >>> is_isbn('9780312498580') # returns true
    >>> is_isbn('1506715214') # returns true

    :param input_string: String to check.
    :param normalize: True to ignore hyphens ("-") in the string (default), false otherwise.
    :return: True if valid ISBN (10 or 13), false otherwise.
    """
    checker = _ISBNChecker(input_string, normalize)
    return checker.is_isbn_13() or checker.is_isbn_10()
